### ISSUE TITLE
capsule proxy with kube oidc proxy

### ISSUE THREAD

**Issue by bbusioc** [clastix/capsule-proxy#307]
Hello,

Is it possible to use capsule-proxy on a cluster that uses for authentication kube oidc proxy?
I've tested capsule in one of our kubernetes clusters (we are just users of the cluster, not admins) and everything is fine.
However we would like to add also capsule proxy in order to allow users to list namespaces.
Unfortunately on this cluster, OIDC authentication is not available so we had to configure kube oidc proxy.

Is it possible to make capsule proxy work on such a cluster and if yes would it be possible to provide some more details on what should be configured or eventually point to the right place for documentation?

Thanks,
Bogdan

**Comment by prometherion** [comment]
Thanks for the question, Bogdan.

We never tested `capsule-proxy` with the `kube-oidc` project, so it's tough for me to understand how it can work.

What I could suggest is to run Capsule Proxy by pointing it to the `kube-oidc-proxy` rather than the Kubernetes API Server: this would require some mangling with the CA and the `kubeconfig`.

If you're encountering any issues with this we could try to move the discussion to the [Capsule Kubernetes Slack channel](https://kubernetes.slack.com/archives/C03GETTJQRL): I guess there could be some sensitive information.

**Comment by bbusioc** [comment]
thanks for the answer.
yes, I thought about this; but I couldn't find any setting in capsule proxy in order to change where capsule proxy is pointing.
do you have an idea how this could be achieved?

**Comment by prometherion** [comment]
You could mount a `kubeconfig` in the Pod and reference it as an environment variable.

Not sure if there's going to be a precedence order between a user-provided kubeconfig and the SA token, tho.

**Comment by bbusioc** [comment]
I managed to make this work.
I'm leaving here what I did, maybe it will help other people in the future.
So, I pointed from kube oidc proxy to go on capsule proxy instead of default kubernetes api.
In order to do this I used the below arguments (in kube oidc proxy):
```bash
            - '--server=https://capsule-proxy.capsule-system.svc:9001'
            - '--insecure-skip-tls-verify'
            - '--token=$(OIDC_SA_TOKEN)'
```

The first 2 I think they are self explanatory, the last one I've added it because for some reason when switching from default kubernetes api to capsul proxy, kube oidc proxy was not adding anymore in the request the Authorization header.
And OIDC_SA_TOKEN is added as an env variable in the pod.

This works, but however I think I might have found a bug.
My user belongs to 2 groups, we will name them group1 and group2.
group1 is owner of a capsule tenant that has 3 namespaces created.
group2 is not an owner of any capsule tenant.
So, from kube oidc proxy to capsule proxy the request looks something like this:

```bash
curl -s -XGET  -H "Impersonate-User: my_user" -H "Impersonate-Group: group1" -H "Impersonate-Group: group2" -H "Authorization: Bearer $TOKEN" 'https://[capsule_url]/api/v1/namespaces'
```

When the request is done with the impersonate groups headers in this order, everything is fine and I can retrieve my 3 namespaces.
If the request is done with the group2 first (which has no tenant), so like this:

```bash
curl -s -XGET  -H "Impersonate-User: my_user" -H "Impersonate-Group: group2" -H "Impersonate-Group: group1" -H "Authorization: Bearer $TOKEN" 'https://[capsule_url]/api/v1/namespaces'
```

I receive an empty list.
It looks like capsule proxy takes into consideration only the first impersonate group header.
To deploy capsule proxy I'm using helm chart version 0.4.9 (app. version 0.4.4).

This behaviour doesn't seem right from my point of view.
What do you think?

**Comment by bbusioc** [comment]
I think this could be a bug.
Should I open a new issue on this?

**Comment by prometherion** [comment]
I've just merged #306 which is expected to solve the issue you're facing, as explained here: https://github.com/clastix/capsule-proxy/pull/306#issuecomment-1740934560

Keep in mind the used `kubeconfig` with user impersonation is straightforward equal to the `curl` with the impersonate headers.

I'm closing this issue since it seems fixed from my perspective, feel free to add further details in case it doesn't.
